### ISSUE TITLE
[FEATURE] Store Statistics and Checkpoints

### ISSUE THREAD

**Issue by madisunleung** [COMP3122-2324-Project/Simpsyber#15]
As the player tackles one or more objectives at each level, the game could strategically implement checkpoints to allow them to resume from specific points rather than forcing a complete restart upon failure. **Currently, every setback results in a loss of progress and statistics, forcing them to exit the game and start anew.** 

For example, I was not able to meet the objectives for Level 3-4. When I failed to meet the objectives and ran out of time, the game popped up the 'Game Over' notification and redirected the player to the starting page. When I re-entered the game without reentering my ID and name, I lost all the statistics that I had in the previous game. 

<img width="1051" alt="Screenshot 2024-02-27 at 2 43 07 AM" src="https://github.com/COMP3122-2324-Project/Simpsyber/assets/92859038/e6619b93-35af-4b3d-9a0f-8a2c3b85a08e">

[
<img width="1053" alt="Screenshot 2024-02-27 at 3 00 47 AM" src="https://github.com/COMP3122-2324-Project/Simpsyber/assets/92859038/c673b030-3296-4ce1-8011-c6695406f648">
](url)

To enhance the game experience. I would like to _propose integrating a function for retaining user statistics within the game framework_. Implementing checkpoints throughout each level will allow the players to resume gameplay from specific points rather than commencing anew after a setback. By storing player statistics, it could not only preserve the game history but also gain valuable insights into the gameplay patterns and strategies. This feature would allow them to reflect on their past performances, identify areas for improvement, and track their progress over time. 

On top of that, it would foster a deeper connection to the game by enabling them to understand and appreciate the journey through the game. In essence, the implementation of stored statistics has the potential to enhance our gaming experiences, offering them a chance to revisit history and evolve as players. What do you think about this idea?

**Comment by jackyngaipan01** [comment]
This sounds like a great idea, and I also have the following suggestions to hopefully make it even better.
Mission-based objectives: Break up the objectives into smaller tasks or missions. If the player fails a mission, they can just retry that part instead of restarting the entire level. This kind of granularity keeps the player engaged.

Tutorial and Practice Modes: Include tutorial levels or practice areas in the game to allow players to learn the mechanics without penalty. This helps them develop skills before facing tougher challenges.

**Comment by ln2-muyuan** [comment]
I agree with you, and the current mode is too difficult for players to easily game over. I think it would be possible to set up a tooltip to tell players what are the important issues right now to improve their score, and set a higher margin for error.

**Comment by RaymondTMP** [comment]
I agree with your idea. Adding checkpoints will encourage players to try more after failure. Currently, I am unable to meet some of the objectives. It takes time to understand the game mechanism. I think it would be possible to generate a detailed summary when the game is over. This would be helpful to let players which part they do wrong.

**Comment by ctszkeung** [comment]
I agree with your point. The current difficulty is not that friendly to newbie especially in the first two weeks. It is easily over budget. So I recommend when the user passes every week, then make a save point. When the game is over, user can choose to start from the week before game over.
